### DISCUSSION TITLE
Automatically refresh the webpage to fetch updated data

### DISCUSSION THREAD

**Discussion by msfjarvis** [glanceapp/glance#104] [General]
I don't know if I'm holding this wrong but I keep a pinned tab with Glance in my browser and it never updates with new data until I manually refresh it. Ideally I would expect the web side of things to poll the internal API every so often but that doesn't seem to happen.

FWIW, my local installation now runs with this naive patch as a workaround: https://github.com/msfjarvis/dotfiles/blob/bf02a54f2a629efde5f933a513b33ecce8a6af51/packages/glance/auto-update.patch

**Comment by svilenmarkov** [comment]
Hey,

Glance doesn't currently have a way of automatically updating the data on the front-end. There's a few reasons for that:

* It seems like bad UX for things to move about on their own. If I notice a post (or any other link/information) that piques my interest right before the page updates and that post disappears, I'd be upset at having to go search for it.
* I dread the days in which I'll have to look for alternative APIs to the ones I'm using now, some of which will inevitably get taken down at some point. It's why I try to play nice by being mindful of their operating costs and have default cache times of around an hour or more. I also have Glance pinned but I might only look at it 5-6 times a day, if it was automatically updating in the background it'd be making many times more requests than it needs to. Accounting for the presumably thousands of people who now use Glance, that results in a potentially hefty increase in traffic for some of these APIs. 
* If pages took 20-30 seconds to load it would make more sense to pre-fetch stuff, but with the current load times of a second or two I really don't see the issue with just hitting refresh. 

Are there ways to reduce the impact of some of these concerns? For sure, though ultimately making Glance a real time dashboard was never a goal.

**Reply by msfjarvis** [reply]
That's completely reasonable, thanks for taking the time to respond in such detail.

**Reply by rubiojr** [reply]
@svilenmarkov would it be useful to make this optional (maybe disabled by default) for custom extensions? I'm currently pulling data from https://www.githubstatus.com/api that Displays GitHub Status in real time in a tablet and it'd be nice if that automatically updates so I can see the Octoverse going :red_circle:, without having to manually refresh.

**Reply by svilenmarkov** [reply]
@rubiojr It would require reworking a bunch of the JS since it currently executes once on content load and does not support unmounting/reloading specific regions. It's not essential, but it does mean that features you may rely on such as relative time or collapsible lists wouldn't work. 

A quick and dirty hack to somewhat achieve this behavior for a specific page (by just refreshing the whole page) would be with the following:

```yaml
branding:
  custom-footer: |
    <script>
    if (pageData.slug === 'home')
      setTimeout(() => location.reload(), 60 * 1000);
    </script>
```

This would automatically refresh the home page every minute. 

When I get to refactoring the front-end code I'll keep this in mind so maybe it could become a simple on/off switch for each widget in the future.

**Reply by rubiojr** [reply]
Thanks for the tip, that's awesome!
